feat(pallet-gear): Add compatibility check for code instrumentation

ukint-vs opened the pull request:
Implement a `try_state` function that verifies all stored program codes can be successfully instrumented with the current schedule before runtime upgrades.

- Iterates through all codes in `OriginalCodeStorage`
- Re-instrument all codes ignoring known incompatibilities

Successor to https://github.com/gear-tech/gear/pull/4598

@gear-tech/dev

ukint-vs committed:
feat(pallet-gear): Add try-state compatibility check for code instrumentation

breathx reviewed (commented):
Except questions from above, looks good

breathx commented on pallets/gear/src/lib.rs:
```diff
@@ -516,6 +520,78 @@ pub mod pallet {
         }
+
+        #[cfg(feature = "try-runtime")]
+        fn try_state(_n: BlockNumberFor<T>) -> Result<(), sp_runtime::TryRuntimeError> {
```
and extend logging please: how many codes checked etc

ukint-vs replied:
> how long does it take to run?
Mainnet CI - 1m 1s master vs 1m 7s try-state branch
> do we run it in CI?
 Yes. Mainnet only
> could you provide errors associated with these codes: like code A is not valid due to Error1, code B due to Error2 etc
done
> you may also iterate using classic wrappers: put it directly in GearProgram pallet or smth like this
I think pallet-gear is more suited for this check

ukint-vs committed:
Add logging

ukint-vs committed:
expect error for known codes

ukint-vs committed:
add reasoning

ukint-vs committed:
Merge branch 'master' into vs/reinstrument-check-try-state

ukint-vs committed:
update args

ukint-vs committed:
impl PartialEq, add new expected errors

ukint-vs committed:
add try-state testnet check

ukint-vs committed:
revert hex dev-dep

ukint-vs committed:
use aria2c to download snapshots

ukint-vs committed:
fix timeout

ukint-vs committed:
rm script

ukint-vs committed:
update workflow

ukint-vs committed:
Add code len check

ukint-vs committed:
maximize disk space

ukint-vs committed:
use /mnt storage

ukint-vs committed:
set swap

ukint-vs committed:
rm disk cleaner

ukint-vs committed:
disk free space

ukint-vs committed:
fix path

ukint-vs committed:
disable heavy tasks

ukint-vs committed:
change container

ukint-vs committed:
add wget

ukint-vs committed:
skip steps

ukint-vs committed:
install binutils

ukint-vs committed:
change container

ukint-vs committed:
rm linker

ukint-vs committed:
set target

ukint-vs committed:
disable name check

ukint-vs committed:
update Eq

ark0f commented on .github/workflows/build-linux.yml:
```diff
@@ -278,12 +278,65 @@ jobs:
-        run: cargo build -p vara-runtime -F std,try-runtime --locked ${{ matrix.profiles.flags }}
+        run: cargo build -p vara-runtime -F std,try-runtime --locked ${{ matrix.profiles.flags }} --target x86_64-unknown-linux-gnu
+
+      - name: "Test: Download Vara Testnet snapshot"
```
Can be turned into matrix instead of Testnet + Mainnet sequential steps

ukint-vs committed:
Merge branch 'master' into vs/reinstrument-check-try-state

ukint-vs committed:
update job

ukint-vs committed:
rm build-linux.yml

ukint-vs committed:
Merge branch 'master' into vs/reinstrument-check-try-state

ukint-vs committed:
testnet try-state check

ukint-vs committed:
free disk space steps

ukint-vs committed:
rm only android dotnet

ukint-vs committed:
rm target dir

ukint-vs committed:
swapfile

ukint-vs committed:
mainnet/testnet matrix

ark0f commented on .github/workflows/build.yml:
```diff
@@ -341,8 +341,19 @@ jobs:
+            snapshot_url: https://snapshots.vara-network.io/vara-mainnet.snap
+            snapshot_file: vara_ci.snap
+            disable_spec_name_check: false
+            run_checks_all: true
```
```suggestion
            args: --checks all
```
No need to run `on-runtime-upgrade` twice

ark0f commented on .github/workflows/build.yml:
```diff
@@ -341,8 +341,19 @@ jobs:
+            run_checks_all: true
       fail-fast: false
-    name: try-runtime (${{ matrix.profiles.name }})
+    name: try-runtime (${{ matrix.profiles.name }} / ${{ matrix.network.name }})
```
Please, use another naming or it looks like this in UI

<img width="353" height="85" alt="Image" src="https://github.com/user-attachments/assets/d4c31c3a-141e-40d2-844c-8b42fd4cc6ac" />

ukint-vs committed:
Merge remote-tracking branch 'origin/master'

ukint-vs committed:
update job

ukint-vs committed:
free disk space

ukint-vs committed:
try state

ukint-vs committed:
fix

ukint-vs committed:
Merge branch 'master' into vs/reinstrument-check-try-state

ukint-vs commented:
Addressed review threads (API reply endpoint returned 404, so posting a summary here):

1. [discussion_r2382152750](https://github.com/gear-tech/gear/pull/4864#discussion_r2382152750): try-state now runs for Vara Testnet and Mainnet via matrix; code IDs map to expected errors; logs include total checked/failed/ignored and per-code errors. No local timing numbers yet.
2. [discussion_r2383963655](https://github.com/gear-tech/gear/pull/4864#discussion_r2383963655): instrumentation attempt uses if let Err in try_state.
3. [discussion_r2383964393](https://github.com/gear-tech/gear/pull/4864#discussion_r2383964393): no mem::discriminant usage; CodeError PartialEq compares inner enums except Validation/Module.
4. [discussion_r2386836413](https://github.com/gear-tech/gear/pull/4864#discussion_r2386836413): CodeError PartialEq returns true for same variant and inner error, except Validation/Module compare by variant only.
5. [discussion_r2386839675](https://github.com/gear-tech/gear/pull/4864#discussion_r2386839675): added code size precheck vs schedule.limits.code_len.
6. [discussion_r2386841474](https://github.com/gear-tech/gear/pull/4864#discussion_r2386841474): ignore list is for known legacy testnet incompatibilities; each ignored code is logged and counted.
7. [discussion_r2466053679](https://github.com/gear-tech/gear/pull/4864#discussion_r2466053679): outdated, build-linux.yml removed after merge from master.
8. [discussion_r2466124880](https://github.com/gear-tech/gear/pull/4864#discussion_r2466124880): outdated, build-linux.yml removed.
9. [discussion_r2466127496](https://github.com/gear-tech/gear/pull/4864#discussion_r2466127496): outdated, try-runtime is now a matrix job in build.yml.
10. [discussion_r2535608635](https://github.com/gear-tech/gear/pull/4864#discussion_r2535608635): single snapshot filename (vara_ci.snap).
11. [discussion_r2535615329](https://github.com/gear-tech/gear/pull/4864#discussion_r2535615329): per-network matrix with try_runtime_args; still run testnet + mainnet, mainnet uses --checks all.
12. [discussion_r2535616266](https://github.com/gear-tech/gear/pull/4864#discussion_r2535616266): args now come from matrix (disable-spec-name-check and checks try-state for testnet; checks all for mainnet).
13. [discussion_r2535619928](https://github.com/gear-tech/gear/pull/4864#discussion_r2535619928): job renamed to try-runtime - <network> (<profile>).
14. [discussion_r2535625660](https://github.com/gear-tech/gear/pull/4864#discussion_r2535625660): merged master; obsolete steps removed/updated in try-runtime job.
15. [discussion_r2535705851](https://github.com/gear-tech/gear/pull/4864#discussion_r2535705851): signature is validated before duplicate; mismatched signatures surface InvalidImportFnSignature.
16. [discussion_r2566074490](https://github.com/gear-tech/gear/pull/4864#discussion_r2566074490): inlined snapshot filename; no extra env var.
17. [discussion_r2566078591](https://github.com/gear-tech/gear/pull/4864#discussion_r2566078591): switched to cargo clean.
18. [discussion_r2566082319](https://github.com/gear-tech/gear/pull/4864#discussion_r2566082319): removed swap enable/disable.
19. [discussion_r2566087360](https://github.com/gear-tech/gear/pull/4864#discussion_r2566087360): done via try_runtime_args to simplify the step.

breathx commented:
Closing for now as it's not a current priority — can be revisited later.
